File: cpp/search/string/RabinKarp.cpp
#include <iostream>
#include <string>
using namespace std;

class RabinKarp {
public:
    int search(string text, string pattern) {
        int code = getHashCode(text, pattern.size());
        int patternCode = getHashCode(pattern, pattern.size());
        for (int i = 0; i <= (int) text.size() - (int) pattern.size(); ++i) {
            if (code == patternCode) {
                int index = i;
                int patternIndex = 0;
                while (index < (int) text.size()) {
                    if (text[index] != pattern[patternIndex]) {
                        break;
                    }

                    if (patternIndex == pattern.size() - 1) {
                        return index - patternIndex;
                    }
                    ++index;
                    ++patternIndex;
                }
            }

            if (i < (int) text.size() - (int) pattern.size()) {
                code += text[i + pattern.size()] - text[i];
            }
        }
        return -1;
    }

private:
    int getHashCode(string text, int length) {
        int code = 0;
        for (int i = 0; i < length; ++i) {
            code += text[i] - 'a';
        }
        return code;
    }
};

File: java/search/string/RabinKarp.java
package search.string;

public class RabinKarp {
    public int search(String text, String pattern) {
        int code = getHashCode(text, pattern.length());
        int patternCode = getHashCode(pattern, pattern.length());
        for (int i = 0; i <= text.length() - pattern.length(); ++i) {
            if (code == patternCode) {
                int index = i;
                int patternIndex = 0;
                while (index < text.length()) {
                    if (text.charAt(index) != pattern.charAt(patternIndex)) {
                        break;
                    }

                    if (patternIndex == pattern.length() - 1) {
                        return index - patternIndex;
                    }
                    ++index;
                    ++patternIndex;
                }
            }

            if (i < text.length() - pattern.length()) {
                code += text.charAt(i + pattern.length()) - text.charAt(i);
            }
        }
        return -1;
    }

    private int getHashCode(String text, int length) {
        int code = 0;
        for (int i = 0; i < length; ++i) {
            code += text.charAt(i) - 'a';
        }
        return code;
    }
}
